packages/vite/src/node/server/middlewares/indexHtml.ts

```diff
@@ -187,48 +189,59 @@ const devHtmlHook: IndexHtmlTransformHook = async (
       server?.moduleGraph.invalidateModule(module)
     }
     s.overwrite(
-      node.loc.start.offset,
-      node.loc.end.offset,
+      node.sourceCodeLocation!.startOffset,
+      node.sourceCodeLocation!.endOffset,
       `<script type="module" src="${modulePath}"></script>`,
       { contentOnly: true }
     )
   }
 
   await traverseHtml(html, htmlPath, (node) => {
-    if (node.type !== NodeTypes.ELEMENT) {
+    if (!nodeIsElement(node)) {
       return
     }
 
     // script tags
-    if (node.tag === 'script') {
-      const { src, isModule } = getScriptInfo(node)
+    if (node.nodeName === 'script') {
+      const { src, sourceCodeLocation, isModule } = getScriptInfo(node)
 
       if (src) {
-        processNodeUrl(src, s, config, htmlPath, originalUrl, moduleGraph)
-      } else if (isModule && node.children.length) {
+        processNodeUrl(
+          src,
+          sourceCodeLocation!,
+          s,
+          config,
+          htmlPath,
+          originalUrl,
+          moduleGraph
+        )
+      } else if (isModule && node.childNodes.length) {
         addInlineModule(node, 'js')
       }
     }
 
-    if (node.tag === 'style' && node.children.length) {
-      const children = node.children[0] as TextNode
+    if (node.nodeName === 'style' && node.childNodes.length) {
+      const children = node.childNodes[0] as DefaultTreeAdapterMap['textNode']
       styleUrl.push({
-        start: children.loc.start.offset,
-        end: children.loc.end.offset,
-        code: children.content
+        start: children.sourceCodeLocation!.startOffset,
+        end: children.sourceCodeLocation!.endOffset,
+        code: children.value
       })
     }
 
     // elements with [href/src] attrs
-    const assetAttrs = assetAttrsConfig[node.tag]
+    const assetAttrs = assetAttrsConfig[node.nodeName]
     if (assetAttrs) {
-      for (const p of node.props) {
-        if (
-          p.type === NodeTypes.ATTRIBUTE &&
-          p.value &&
-          assetAttrs.includes(p.name)
-        ) {
-          processNodeUrl(p, s, config, htmlPath, originalUrl)
+      for (const p of node.attrs) {
+        if (p.value && assetAttrs.includes(p.name)) {
```

Comment:
This part needs to consider duplicate attributes, too. (I will still work so it's not necessary)